### ISSUE TITLE
Tutorial: Mongorepo with typescript, how to benefit from tsbuildcache info

### ISSUE THREAD

**Issue by simllll** [actions/cache#459]
We are using github actinos to run our tests, all works fine, except it takes ages to compile everything. Therefore we were looking into caching artifacts and quickly implemented actions/cache (y).
But then it turned out, even though files are pre-compiled, nothing has changed and the buildcache info file is valid, typecript still recompiles everything.
As it turned out, github checkout (even with fetch-detph: 0) is not restoring last modified time of a file. Therefore typescript always assumes that the buildcache is invalid, even if you just compiled in in a job before that.

Therefore I like to write down how we have solved this:
1.) you need to ensure last modified time is the same
2.) ensure that .tsbuildinfo and build artifacts are included in the cache
3.) ensure that changed files are rebuild, even the buildinfo cache is still "valid".

What we did:
1.) detect what has changed (we need this later on)
```
- uses: dorny/paths-filter@v2
        id: filter
        with:
          list-files: 'json'
          filters: |
            packages: // <-- specify all your monorepo locations (usually you probabaly just have one)
              - 'packages/**'
            infrastructure:
              - 'infrastructure/**'
            services:
              - 'services/**'
```
2.) restore modified time of files via git-restore-mtime
```
      - name: Restore mtime for git checkout
        run: sudo apt install git-restore-mtime && sudo curl -o /usr/lib/git-core/git-restore-mtime https://raw.githubusercontent.com/MestreLion/git-tools/v2020.09/git-restore-mtime && git restore-mtime

```

3.) set up cache (include node_modules, build artifacts and tsbuildinfo!)
```
- name: Setup cache for build and dependencies
        uses: actions/cache@v2
        id: cache
        with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/.tsbuildinfo
            ~/.npm
          key: ....
```

4.)  ensure buildcache is always in a valid state, therefore remove tsbuildinfo files from directories that have changed.
  
```
  - name: Remove buildcache infos
        uses: hokify/remove-buildcache-action@v1
        with:
          changed-files: ${{ steps.filter.outputs.packages_files }}
          tsBuildInfoFile: '.tsbuildinfo'
```

Hope this helps, it's just a very plain description, but I was looking for this for ages, and couldn't find anything.

**Comment by simllll** [comment]
I put up a README.md file here: https://github.com/hokify/remove-buildcache-action/tree/master

**Comment by phil-lgr** [comment]
Thank you for this @simllll 

I'm wondering what did you use as key

```yaml
- name: Setup cache for build and dependencies
        uses: actions/cache@v2
        id: cache
        with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/.tsbuildinfo
            ~/.npm
          key: ....
```

since there are multiple tsbuildinfo files...

**Comment by phil-lgr** [comment]
Something like 

```yaml
    - name: Setup cache for build and dependencies
      uses: actions/cache@v2
      id: cache
      with:
        path: |
          node_modules
          */*/node_modules
          */*/lib
          */*/.tsbuildinfo
        key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}-${{ hashFiles('**/.tsbuildinfo') }}
```
?

**Comment by simllll** [comment]
Hi @phil-lgr ,No but we cache these files

**Comment by phil-lgr** [comment]
@simllll ok thx..

could you share what key you used ?

I'm trying

```yaml
key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}-${{ hashFiles('**/*.tsbuildinfo') }}
```

**Comment by phil-lgr** [comment]
do you have a single `.tsbuildinfo` at the root ?

In my case I'm using composite project so have 1x per package

**Comment by phil-lgr** [comment]
I guess I'd have to modify `hokify/remove-buildcache-action`

```yaml
    #
    # Remove buildcache infos if src files have changed
    #
    - name: Remove buildcache infos
      uses: hokify/remove-buildcache-action@v1
      with:
        changed-files: ${{ steps.filter.outputs.packages_files }}
        tsBuildInfoFile: '.tsbuildinfo'
```

To handle multiple tsbuildinfo

**Comment by simllll** [comment]
our cache key is this:
```
 with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/buildcache.tsbuildinfo
            ~/.npm
          key: ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-
```

we also have a monorepo with composite mode, should work with this already :)

**Comment by simllll** [comment]
If it helps, I share the whole github action yml file with you:

* we have a monorepo with several folders (packages, services, ..)

```
name: Run test pipeline

on:
  workflow_dispatch:
#  push:
#    branches:
#      - master
  pull_request:
    types:
      - assigned
      - opened
      - synchronize
      - reopened
      - ready_for_review
    branches:
      - master

env:
  NODE_VERSION: '15'
  # codecov disabled for now, as we have not enough test to have a benefit from it
  ENABLE_CODECOV: false

jobs:
  prebuild:
    if: github.event.pull_request.draft == false
    name: build core packages
    runs-on: ubuntu-latest

    outputs:
      packages: ${{ steps.filter.outputs.packages }}
      packages_files: ${{ steps.filter.outputs.packages_files }}
      infrastructure: ${{ steps.filter.outputs.infrastructure }}
      infrastructure_files: ${{ steps.filter.outputs.infrastructure_files }}
      services: ${{ steps.filter.outputs.services }}
      services_files: ${{ steps.filter.outputs.services_files }}

    steps:
      - name: Cancel Previous Runs
        uses: styfle/cancel-workflow-action@0.8.0
        with:
          access_token: ${{ github.token }}
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - uses: dorny/paths-filter@v2
        id: filter
        with:
          list-files: 'json'
          filters: |
            packages:
              - 'packages/**'
            infrastructure:
              - 'infrastructure/**'
            services:
              - 'services/**'
      - name: Restore mtime for git checkout
        run: sudo apt install git-restore-mtime && sudo curl -o /usr/lib/git-core/git-restore-mtime https://raw.githubusercontent.com/MestreLion/git-tools/v2020.09/git-restore-mtime && git restore-mtime
      - name: Setup cache for build and dependencies
        uses: actions/cache@v2
        id: cache
        with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/buildcache.tsbuildinfo
            ~/.npm
          key: ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-
      - name: Install Node.js
        if: steps.cache.outputs.cache-hit != 'true'
        uses: actions/setup-node@v2.1.4
        with:
          node-version: ${{ env.NODE_VERSION }}
      - name: Setup node config
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm config set unsafe-perm true
      - name: Install project dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci --no-audit --prefer-offline
        env:
          PUPPETEER_SKIP_CHROMIUM_DOWNLOAD: true
          HUSKY_SKIP_INSTALL: true
      - name: Remove buildcache infos
        if: steps.cache.outputs.cache-hit != 'true'
        uses: hokify/remove-buildcache-action@v1
        continue-on-error: true
        with:
          changed-files: ${{ steps.filter.outputs.packages_files }}
          tsBuildInfoFile: 'buildcache.tsbuildinfo'
      - name: Build
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm run build
  build-infrastructure:
    name: Build infrastructure and run tests
    #    runs-on: self-hosted
    runs-on: ubuntu-latest
    needs: prebuild
    if: ${{ needs.prebuild.outputs.infrastructure == 'true' }}

    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Restore mtime for git checkout
        run: sudo apt install git-restore-mtime && sudo curl -o /usr/lib/git-core/git-restore-mtime https://raw.githubusercontent.com/MestreLion/git-tools/v2020.09/git-restore-mtime && git restore-mtime
      - name: Setup cache for build and dependencies
        uses: actions/cache@v2
        with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/buildcache.tsbuildinfo
            ~/.npm
          key: ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-
      - name: Install Node.js
        uses: actions/setup-node@v2.1.4
        with:
          node-version: ${{ env.NODE_VERSION }}
      - name: Remove buildcache infos
        uses: hokify/remove-buildcache-action@v1
        continue-on-error: true
        with:
          changed-files: ${{ needs.prebuild.outputs.infrastructure_files }}
          tsBuildInfoFile: 'buildcache.tsbuildinfo'
      - name: Build infrastructure
        run: npm run build:infrastructure
      - name: Run tests
        run: npm run coverage:infrastructure
      - name: Upload coverage to Codecov
        if: ${{ env.ENABLE_CODECOV == 'true' }}
        uses: codecov/codecov-action@v1
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          flags: infrastructure
          fail_ci_if_error: false
  build-packages:
    name: Build packages and run tests
    #    runs-on: self-hosted
    runs-on: ubuntu-latest
    needs: prebuild
    if: ${{ needs.prebuild.outputs.packages == 'true' }}

    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Restore mtime for git checkout
        run: sudo apt install git-restore-mtime && sudo curl -o /usr/lib/git-core/git-restore-mtime https://raw.githubusercontent.com/MestreLion/git-tools/v2020.09/git-restore-mtime && git restore-mtime
      - name: Setup cache for build and dependencies
        uses: actions/cache@v2
        with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/buildcache.tsbuildinfo
            ~/.npm
          key: ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-
      - name: Install Node.js
        uses: actions/setup-node@v2.1.4
        with:
          node-version: ${{ env.NODE_VERSION }}
      #      - name: Install environment dependencies
      #        run: sudo apt-get update && sudo apt-get install -y gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libgbm-dev
      - name: Remove buildcache infos
        uses: hokify/remove-buildcache-action@v1
        continue-on-error: true
        with:
          changed-files: ${{ needs.prebuild.outputs.packages_files }}
          tsBuildInfoFile: 'buildcache.tsbuildinfo'
      - name: Build packages
        run: npm run build:packages
      - name: Run tests
        run: npm run coverage:packages
      - name: Upload coverage to Codecov
        if: ${{ env.ENABLE_CODECOV == 'true' }}
        uses: codecov/codecov-action@v1
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          flags: packages
          fail_ci_if_error: false
  build-services:
    name: Build services and run tests
    #    runs-on: self-hosted
    runs-on: ubuntu-latest
    needs: prebuild
    if: ${{ needs.prebuild.outputs.services == 'true' }}

    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - name: Restore mtime for git checkout
        run: sudo apt install git-restore-mtime && sudo curl -o /usr/lib/git-core/git-restore-mtime https://raw.githubusercontent.com/MestreLion/git-tools/v2020.09/git-restore-mtime && git restore-mtime
      - name: Setup cache for build and dependencies
        uses: actions/cache@v2
        with:
          path: |
            node_modules
            */*/node_modules
            */*/dist
            */*/buildcache.tsbuildinfo
            ~/.npm
          key: ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-${{ hashFiles('package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-${{ hashFiles('packages/*/src/**/*') }}-
            ${{ runner.os }}-${{ hashFiles('cache_root_version') }}-node-build-
      - name: Install Node.js
        uses: actions/setup-node@v2.1.4
        with:
          node-version: ${{ env.NODE_VERSION }}
      - name: Remove buildcache infos
        uses: hokify/remove-buildcache-action@v1
        continue-on-error: true
        with:
          changed-files: ${{ needs.prebuild.outputs.services_files }}
          tsBuildInfoFile: 'buildcache.tsbuildinfo'
      - name: Build services
        run: npm run build:services
      - name: Run tests
        run: npm run coverage:services
      - name: Upload coverage to Codecov
        if: ${{ env.ENABLE_CODECOV == 'true' }}
        uses: codecov/codecov-action@v1
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          flags: services
          fail_ci_if_error: true

```

**Comment by phil-lgr** [comment]
@simllll thanks for sharing

BTW it's crazy that we can't reuse steps on github action see https://github.com/actions/runner/issues/646

TL;DR you can create composite actions, but `uses` isn't supported right now, it sucks because we have to copy the steps for each job :/
